Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch

```diff
@@ -0,0 +1,294 @@
+ 
+     public long h() {
+         long i = 0L;
+@@ -183,6 +184,7 @@ public class PacketDataSerializer extends ByteBuf {
+     public UUID i() {
+         return new UUID(this.readLong(), this.readLong());
+     }
++    public UUID readUuid() { return this.i(); } // Paper - OBFHELPER
+ 
+     public PacketDataSerializer d(int i) {
+         while ((i & -128) != 0) {
+@@ -316,6 +318,7 @@ public class PacketDataSerializer extends ByteBuf {
+             }
+         }
+     }
++    public String readUtf(int maxLength) { return this.e(maxLength); } // Paper - OBFHELPER
+ 
+     public PacketDataSerializer a(String s) {
+         return this.a(s, 32767);
+diff --git a/src/main/java/net/minecraft/server/PacketLoginInCustomPayload.java b/src/main/java/net/minecraft/server/PacketLoginInCustomPayload.java
+index e3c0094f7..edfd4a506 100644
+--- a/src/main/java/net/minecraft/server/PacketLoginInCustomPayload.java
++++ b/src/main/java/net/minecraft/server/PacketLoginInCustomPayload.java
+@@ -3,8 +3,8 @@ package net.minecraft.server;
+ import java.io.IOException;
+ 
+ public class PacketLoginInCustomPayload implements Packet<PacketLoginInListener> {
+-    private int a;
+-    private PacketDataSerializer b;
++    private int a; public int getId() { return a; } // Paper - OBFHELPER
++    private PacketDataSerializer b; public PacketDataSerializer getBuf() { return b; } // Paper - OBFHELPER
+ 
+     public PacketLoginInCustomPayload() {
+     }
+diff --git a/src/main/java/net/minecraft/server/PacketLoginOutCustomPayload.java b/src/main/java/net/minecraft/server/PacketLoginOutCustomPayload.java
+index 9c5559ece..9de0421bb 100644
+--- a/src/main/java/net/minecraft/server/PacketLoginOutCustomPayload.java
++++ b/src/main/java/net/minecraft/server/PacketLoginOutCustomPayload.java
+@@ -10,6 +10,14 @@ public class PacketLoginOutCustomPayload implements Packet<PacketLoginOutListene
+     public PacketLoginOutCustomPayload() {
+     }
+ 
++    // Paper start
++    public PacketLoginOutCustomPayload(int id, MinecraftKey channel, PacketDataSerializer buf) {
++        this.a = id;
++        this.b = channel;
++        this.c = buf;
++    }
++    // Paper end
++
+     public void a(PacketDataSerializer packetdataserializer) throws IOException {
+         this.a = packetdataserializer.g();
+         this.b = packetdataserializer.l();
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index 19f2f2bbd..e723163f2 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -1426,7 +1426,8 @@ public final class CraftServer implements Server {
+         GameProfile profile;
+         // Only fetch an online UUID in online mode
+         if (MinecraftServer.getServer().getOnlineMode()
+-                || (org.spigotmc.SpigotConfig.bungee && com.destroystokyo.paper.PaperConfig.bungeeOnlineMode)) {
```

Comment:
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()